src/tactic/core.lean

```diff
@@ -529,6 +529,66 @@ do g₀ :: _ ← get_goals,
 
 meta def triv' : tactic unit := do c ← mk_const `trivial, exact c reducible
 
+meta def reorder_goals (gs : list (bool × expr)) : new_goals → list expr
+| new_goals.non_dep_first := (gs.reverse.filter $ coe ∘ bnot ∘ prod.fst).map prod.snd ++ (gs.reverse.filter $ coe ∘ prod.fst).map prod.snd
+| new_goals.non_dep_only := (gs.reverse.filter (coe ∘ bnot ∘ prod.fst)).map prod.snd
+| new_goals.all := gs.reverse.map prod.snd
+
+meta def retry_apply_aux : Π (e : expr) (cfg : apply_cfg), list (bool × expr) → tactic (list (name × expr))
+| e cfg gs :=
+focus1 (do {
+     r ← apply_core e cfg,
+     gs' ← get_goals,
+     set_goals (gs' ++ reorder_goals gs cfg.new_goals),
+     return r }) <|>
+do t ← infer_type e >>= whnf,
+   if t.is_pi
+     then do v ← mk_meta_var t.binding_domain,
+             let b := t.binding_body.has_var,
+             e ← head_beta $ e v,
+             retry_apply_aux e cfg ((b, v) :: gs)
+     else apply_core e cfg
```

Comment:
I'm confused: if you reach line 550, haven't you already failed doing exactly the same thing on line 540? Can't we replace line 550 with just `else failed`?

Reply:
That's right. But by finishing up with `apply_core`, I get the error message to be the same as that of `apply`.

Reply:
Ah, smart. A comment ```-- fail with the error message of `apply_core` ``` might be helpful.